modules/shortcodes.php

```diff
@@ -118,12 +118,51 @@ function jetpack_preg_replace_callback_outside_tags( $pattern, $callback, $conte
 	
 	$textarr = wp_html_split( $content );
 	unset( $content );
+
+	// Flag to check we're inside an HTML tag.
+	$inside_tag = '';
+
 	foreach( $textarr as &$element ) {
-	    if ( '' === $element || '<' === $element{0} )
+
+		// If it's an empty element, continue.
+	    if ( '' === $element ) {
 	        continue;
-	    $element = preg_replace_callback( $pattern, $callback, $element );
+		}
+
+		// If this is an opening HTML tag, save it and move to the next element.
+		if ( '<' === $element[0] && '/>' !== substr( $element, -2 ) ) {
+			$inside_tag = $element;
+			continue;
+		}
+
+		// If this is a closing HTML tag, erase the previously saved element.
+		if ( '</' === substr( $element, 0, 2 ) ) {
+			$inside_tag = '';
+			continue;
+		}
+
+		// We're inside an HTML tag.
+		if ( ! empty( $inside_tag ) ) {
+
+			// If the callback it's not in the blacklist, run it.
+			switch ( $callback ) {
+				case 'vimeo_link_callback':
+					// If we're inside a link, don't embed the video.
+					// Covers cases like "Here's a link to the video: <a href="https://vimeo.com/123456">https://vimeo.com/123456</a>".
+					if ( ! preg_match( '/<a\s[^>]+>/i', $inside_tag ) ) {
+						$element = preg_replace_callback( $pattern, $callback, $element );
+					}
+					break;
+
+				default:
+					$element = preg_replace_callback( $pattern, $callback, $element );
+					break;
+			}
+
+		} else {
+			$element = preg_replace_callback( $pattern, $callback, $element );
+		}
```

Comment:
I'm trying to figure out when we _do_ and when we _don't_ run the exact same code here at varying indention levels. Is it possible that the only place we don't run this code is when we're inside an anchor tag?

```php
if ( ! empty( $inside_tag ) && 'vimeo_link_callback' === $callback && isTag( 'a', $inside_tag ) ) {
	continue;
}

$element = preg_replace_callback( $pattern, $callback, $element );
```

Of course, we still have some ugliness here. If this is the only exception we ever have, it's probably fine, but if we start adding more rules, why not separate it a bit further still…

```php
function skipVimeoLinksInAnchorTags( $element, $callback, $inside_tag ) {
	return (
		! empty( $inside_tag ) &&
		'vimeo_link_callback' === $callback &&
		isTag( 'a', $inside_tag )
	);
}

…

if (
	skipVimeoLinksInAnchorTags( $element, $callback, $inside_tag ) ||
	skipAnythingInsideNastyTags( $inside_tag ) ||
	skipHttpsLinksForYoutube( $element ) ||
	skipRandomStuffs()
) { continue; }

$element = preg_replace_callback( $pattern, $callback, $element );
```